File: stormbird/src/line_force_model/input_power.rs
// Copyright (C) 2024, NTNU
// Author: Jarle Vinje Kramer <jarlekramer@gmail.com; jarle.a.kramer@ntnu.no>
// License: GPL v3.0 (see separate file LICENSE or https://www.gnu.org/licenses/gpl-3.0.html)

use serde::{Serialize, Deserialize};

use stormath::type_aliases::Float;
use stormath::interpolation::linear_interpolation;

use crate::line_force_model::span_line::SpanLine;

use stormath::spatial_vector::SpatialVector;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct InputPowerData {
    pub section_models_internal_state_data: Vec<Float>,
    pub input_power_coefficient_data: Vec<Float>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
/// An empirical model to calculate the input power required for driving a wind propulsion device.
///
/// It comes with different modes. Each mode represents different ways of calculating the data.
pub enum InputPowerModel {
    /// Default value. Represent a case where a sail does not need power at all
    NoPower,
    /// Calculates the power using the internal state of the sectional model alone. This could, for
    /// instance, be a power model where the power is calculated directly from the RPS of a rotor
    /// sail
    InterpolateFromInternalState(InputPowerData),
    /// Calculates the power using the the internal state as an aerodynamic power power_coefficient.
    /// 
    /// This then represents a model where the internal state is a measure of the input power, made 
    /// non-dimensional by the density, sail area, and the velocity to the third power.
    FromInternalStateAsPowerCoefficient,
}

impl Default for InputPowerModel {
    fn default() -> Self {
        InputPowerModel::NoPower
    }
}

impl InputPowerModel {
    pub fn input_power_coefficient(&self, section_model_internal_state: Float) -> Float {
        match self {
            InputPowerModel::NoPower => 0.0,
            InputPowerModel::InterpolateFromInternalState(data) => {
                linear_interpolation(
                    section_model_internal_state.abs(),
                    &data.section_models_internal_state_data,
                    &data.input_power_coefficient_data,
                )
                },
            InputPowerModel::FromInternalStateAsPowerCoefficient => {
                section_model_internal_state.abs()
            },
        }
    }

    /// The input power on a given strip, represented by a span lien and chord length.
    pub fn input_power_for_strip(
        &self,
        section_model_internal_state: Float,
        span_line: SpanLine,
        chord_length: Float,
        density: Float,
        velocity: SpatialVector
    ) -> Float {
        let power_coefficient = self.input_power_coefficient(section_model_internal_state);

        match self {
            InputPowerModel::NoPower => 0.0,
            InputPowerModel::InterpolateFromInternalState(_) => {
                power_coefficient * chord_length * span_line.length()
            },
            InputPowerModel::FromInternalStateAsPowerCoefficient => {                
                let dynamic_pressure = 0.5 * density * velocity.length_squared();
                
                let strip_area = chord_length * span_line.length();
                
                power_coefficient * dynamic_pressure * strip_area * velocity.length()
            },
        }
    }
}

File: interfaces/stormbird_setup/src/stormbird_setup/direct_setup/input_power.py
'''
Copyright (C) 2024, NTNU
Author: Jarle Vinje Kramer <jarlekramer@gmail.com; jarle.a.kramer@ntnu.no>
License: GPL v3.0 (see separate file LICENSE or https://www.gnu.org/licenses/gpl-3.0.html)
'''

from ..base_model import StormbirdSetupBaseModel

from enum import Enum

from pydantic import model_serializer

import numpy as np

class InputPowerData(StormbirdSetupBaseModel):
    section_models_internal_state_data: list[float]
    input_power_coefficient_data: list[float]

class InputPowerDataType(Enum):
    NoPower = "NoPower"
    FromInternalStateAlone = "FromInternalStateAlone"
    FromInternalStateAndVelocity = "FromInternalStateAndVelocity"

class InputPowerModel(StormbirdSetupBaseModel):
    '''
    Interface to the input power model
    '''
    input_power_type: InputPowerDataType = InputPowerDataType.NoPower
    input_power_data: InputPowerData | None = None

    @classmethod
    def new_polynomial_rotor_sail_model(
        cls,
        max_power: float,
        max_rps: float,
        area: float
    ) -> "InputPowerModel":
        '''
        Simple model for the power based on a polynomial relationship between the power and RPS.

        The polynomial power is set to 2.5, which comes from data fitted to data from the SWOPP
        project. However, the actual power is scaled based on the supplied values for max_power
        and max_rps.
        '''

        section_models_internal_state_data = np.linspace(0, max_rps, 20)

        power = 2.5

        factor = max_power / (max_rps**power * area)

        input_power_coefficient_data = factor * (section_models_internal_state_data**power)

        return cls(
            input_power_type = InputPowerDataType.FromInternalStateAlone,
            input_power_data = InputPowerData(
                section_models_internal_state_data = section_models_internal_state_data.tolist(),
                input_power_coefficient_data = input_power_coefficient_data.tolist()
            )
        )

    @model_serializer
    def ser_model(self) -> dict[str, object] | str:
        match self.input_power_type:
            case InputPowerDataType.NoPower:
                return self.input_power_type.value
            case InputPowerDataType.FromInternalStateAlone:
                if self.input_power_data is None:
                    raise ValueError("input_power_data must be set for FromInternalStateAlone")

                return {
                    "FromInternalStateAlone": self.input_power_data.model_dump()
                }
            case InputPowerDataType.FromInternalStateAndVelocity:
                if self.input_power_data is None:
                    raise ValueError("input_power_data must be set for FromInternalStateAndVelocity")

                return {
                    "FromInternalStateAndVelocity": self.input_power_data.model_dump()
                }
